zetaclient/chains/bitcoin/observer/event.go

```diff
@@ -0,0 +1,233 @@
+			return errors.Wrap(err, "standard memo contains improper data")
+		}
+
+		// NoAssetCall will be disabled for Bitcoin until full V2 support
+		if memoStd.OpCode == memo.OpCodeCall {
+			return errors.New("NoAssetCall is disabled")
+		}
+
+		// ensure the revert address is valid and supported
+		revertAddress := memoStd.RevertOptions.RevertAddress
+		if revertAddress != "" {
+			btcAddress, err := chains.DecodeBtcAddress(revertAddress, ob.Chain().ChainId)
+			if err != nil {
+				return errors.Wrapf(err, "invalid revert address in memo: %s", revertAddress)
+			}
+			if !chains.IsBtcAddressSupported(btcAddress) {
+				return fmt.Errorf("unsupported revert address in memo: %s", revertAddress)
+			}
+		}
+		event.MemoStd = memoStd
+		receiver = memoStd.Receiver
+	default:
+		// fallback to legacy memo if standard memo decoding fails
+		parsedAddress, _, err := memo.DecodeLegacyMemoHex(hex.EncodeToString(event.MemoBytes))
+		if err != nil { // never happens
+			return errors.Wrap(err, "invalid legacy memo")
+		}
+		receiver = parsedAddress
+	}
+
+	// ensure the receiver is valid
+	if crypto.IsEmptyAddress(receiver) {
+		return errors.New("got empty receiver address from memo")
+	}
+	event.ToAddress = receiver.Hex()
+
+	return nil
+}
+
+// CheckEventProcessability checks if the inbound event is processable
+func (ob *Observer) CheckEventProcessability(event *BTCInboundEvent) bool {
+	// check if the event is processable
+	result := event.CheckProcessability()
+	switch result {
+	case InboundProcessabilityGood:
+		return true
+	case InboundProcessabilityDonation:
+		logFields := map[string]any{
+			logs.FieldChain: ob.Chain().ChainId,
+			logs.FieldTx:    event.TxHash,
+		}
+		ob.Logger().Inbound.Info().Fields(logFields).Msgf("thank you rich folk for your donation!")
+		return false
+	case InboundProcessabilityComplianceViolation:
+		compliance.PrintComplianceLog(ob.logger.Inbound, ob.logger.Compliance,
+			false, ob.Chain().ChainId, event.TxHash, event.FromAddress, event.ToAddress, "BTC")
+		return false
+	}
+
+	return true
+}
+
+// NewInboundVoteV1 creates a MsgVoteInbound message for inbound that uses legacy memo
+func (ob *Observer) NewInboundVoteV1(event *BTCInboundEvent, amountSats *big.Int) *crosschaintypes.MsgVoteInbound {
+	message := hex.EncodeToString(event.MemoBytes)
+
+	return crosschaintypes.NewMsgVoteInbound(
+		ob.ZetacoreClient().GetKeys().GetOperatorAddress().String(),
+		event.FromAddress,
+		ob.Chain().ChainId,
+		event.FromAddress,
+		event.ToAddress,
+		ob.ZetacoreClient().Chain().ChainId,
+		cosmosmath.NewUintFromBigInt(amountSats),
+		message,
+		event.TxHash,
+		event.BlockNumber,
+		0,
+		coin.CoinType_Gas,
+		"",
+		0,
+		crosschaintypes.ProtocolContractVersion_V1,
+		false, // not relevant for v1
+	)
+}
+
+// NewInboundVoteMemoStd creates a MsgVoteInbound message for inbound that uses standard memo
+// TODO: rename this function as 'EventToInboundVoteV2' to use ProtocolContractVersion_V2 and enable more options
```

Comment:
> TODO: rename this function as 'EventToInboundVoteV2'
No this would simply be the default logic

Reply:
updated comment: https://github.com/zeta-chain/node/pull/3025/commits/7a2880bf21e47dfa10d0f3a9dd3756064237afbd